### PR TITLE
Warning and implicit delegate on incomplete implementation.

### PR DESCRIPTION
**Pull request by am-kantox**

### CODE REVIEW

File: lib/dry/behaviour/black_tie.rb [hunk 1]
```diff
@@ -30,7 +29,7 @@ def defprotocol
         singleton_class.send :define_method, method do |receiver, *args|
           receiver.class.ancestors.lazy.map do |c|
             BlackTie.implementations[self].fetch(c, nil)
-          end.reject(&:nil?).first[method].(receiver, *args)
+          end.reject(&:nil?).first[method].(*args.unshift(receiver))
```
Comment (rewritten) [inline lib/dry/behaviour/black_tie.rb]: Warning. You are mutating `args`

File: lib/dry/behaviour/black_tie.rb [hunk 2]
```diff
@@ -39,26 +38,34 @@ def defmethod(name, *params)
       BlackTie.protocols[self][name] = params
     end
 
+    DELEGATE_METHOD = lambda do |klazz, (source, target)|
+      klazz.class_eval do
+        define_method source do |this, *args, **params, &λ|
+          case
+          when !args.empty? && !params.empty? then this.send(target, *args, **params, &λ)
```
Comment (rewritten) [inline lib/dry/behaviour/black_tie.rb]: This clause should cover all cases.

File: lib/dry/behaviour/black_tie.rb [hunk 3]
```diff
@@ -39,26 +38,34 @@ def defmethod(name, *params)
...
-              end
-            end
-          end
-        end
+        mds.each(&DELEGATE_METHOD.curry[singleton_class])
         singleton_class.class_eval(&Proc.new) if block_given? # block takes precedence
       end.tap do |mod|
-        mod.methods(false).each do |m|
+        mod.methods(false).tap do |meths|
+          (BlackTie.protocols[protocol || self].keys - meths).each_with_object(meths) do |m, acc|
```
Comment (rewritten) [inline lib/dry/behaviour/black_tie.rb]: Each_with_object does not yield the accumulator

File: lib/dry/behaviour/black_tie.rb (hunk 1 above)
Comment (am-kantox) [inline lib/dry/behaviour/black_tie.rb]: Right before they are GC’ed.

File: lib/dry/behaviour/black_tie.rb (hunk 3 above)
Comment (am-kantox) [inline lib/dry/behaviour/black_tie.rb]: Disregarded.

**Pull request merged**